src/testRunner/unittests/helpers/tsserver.ts

```diff
@@ -67,6 +68,130 @@ export function patchHostTimeouts(
     }
 }
 
+function patchSessionToHandleWatchEvents(session: TestSession) {
+    const event = session.event;
+    const idToClose = new Map<number, () => void>();
+    session.event = (data, eventName) => {
+        event.call(session, data, eventName);
+        switch (eventName) {
+            case ts.server.CreateFileWatcherEvent:
+                watchFile(data as ts.server.protocol.CreateFileWatcherEventBody);
+                break;
+            case ts.server.CreateDirectoryWatcherEvent:
+                watchDirectory(data as ts.server.protocol.CreateDirectoryWatcherEventBody);
+                break;
+            case ts.server.CloseFileWatcherEvent:
+                closeWatcher(data as ts.server.protocol.CloseFileWatcherEventBody);
+                break;
+            default:
+                break;
+        }
+    };
+
+    function watchFile(event: ts.server.protocol.CreateFileWatcherEventBody) {
+        createWatcher(
+            "watchFile",
+            event,
+            recordChange =>
+                session.host.watchUtils.pollingWatch(
+                    session.host.toNormalizedAbsolutePath(event.path),
+                    {
+                        cb: (fileName, eventKind) =>
+                            recordChange(
+                                event.id,
+                                session.host.windowsStyleRoot ?
+                                    path.win32.resolve(fileName) :
+                                    path.posix.resolve(fileName),

```

Comment:
Is this path just used for debug logging? Surprised to see node's real path lib being used here, but tests do pass...

Reply:
Ye this is testing part - where clients like vscode will send win32 like paths instead of normalized so this is testing only

Reply:
Ah, okay; I think I was just worried about this resolving via a real path on disk like cwd.